### ISSUE TITLE
Request your motions

### ISSUE THREAD

**Issue by godbout** [godbout/kindaVim.docs#139]

**Comment by ylabbe** [comment]
I often use `d$`

**Comment by godbout** [comment]
> I often use `d$`

`d$` and its synonyms (`D` and `dEnd`) are implemented for both the `Accessibility Strategy` and the `Keyboard Strategy`.

more coming...

**Comment by godbout** [comment]
demo on both the `AS` and the `KS`:

https://user-images.githubusercontent.com/121373/200562327-b652ad19-7868-4324-b7e4-2897e6fb07f8.mp4

is it possible maybe that you're using kV on an app that is restricted, and you need to add this app to the `PG-R` Family in the Settings?

checkout the kV menu icon. there's an option to send the current app to The Wizard™ and see if you that app needs some special treatment. if the app is not recognized you can request an investigation and i'll have a look.

**Comment by tddschn** [comment]
I really miss `c-w` in insert mode that deletes the word before the cursor, would you consider adding it?

**Comment by godbout** [comment]
> I really miss `c-w` in insert mode that deletes the word before the cursor, would you consider adding it?

ha. well. took me a while to find the doc for that one ([here](https://vimdoc.sourceforge.net/htmldoc/insert.html#i_CTRL-W)). also, never built a move that works in **INSERT MODE**. because basically this is stealing macOS' standard shortcuts and behaviors. (also i personally never use those moves.)

hmm, so, thinking out loud:
1. is this like `db/cb`, but when in `Insert Mode`?
2. because this is stealing macOS' standard behavior, i'd have to add it as a Setting
3. also because kV has been built to handle keystrokes **ONCE** in Normal Mode, that's gonna require quite some changes. so not sure how long it would take

**Comment by tddschn** [comment]
Thank you for your reply!


> 1. is this like `db/cb`, but when in `Insert Mode`? 

Yes.

> 2. because this is stealing macOS' standard behavior, i'd have to add it as a Setting

That'd be awesome!

> 3. also because kV has been built to handle keystrokes **ONCE** in Normal Mode, that's gonna require quite some changes. so not sure how long it would take

I understand that it's not that easy to implement, please take your time - and it's perfectly okay if you decide not to add this feature. :)

**Comment by godbout** [comment]
i see. got it. thanks for clarifying.

**Comment by godbout** [comment]
https://github.com/godbout/kindaVim.docs/assets/121373/f37ed980-a08d-45ac-a49a-a3f086ef887b

am i missing something? 🤔️

**Comment by itaysk** [comment]
> am I missing something? 🤔️

for some reason it doesn't work for me so I assumed it wasn't supported but I see now in the doc that it should be. 
I will create a separate bug issue if needed. Thanks.

**Comment by godbout** [comment]
> sorry, I was testing in Safari which apparently [isn't not supposed to work OOB (https://github.com/godbout/kindaVim.docs/issues/103). 

OH. yes. after answering i wondered also if you were on an app where kV has to use the Keyboard Strategy.

> FWIW I think the docs about families and limitation could be improved, but that's outside the scope of this issue. Thanks!

nice. anything specific? i may need a review yes. if you have some details about what you think is not good, or is missing, or is confusing etc. please lemme know. that may help. i'm trying to keep the learning curve of kV as flat as possible, but it's also a project that is abusing the AX Framework to do something it's not supposed to do 😂️😂️😂️ and there is indeed limitations. it's a bit hard to explain right from the start, and i'd hoped that people start using kV and realize quite quickly what the strategies and The Wizard are about. but i may be wrong yeah.

**Comment by prurph** [comment]
I'd love to see the ci", ci', ci[, etc motions!

**Comment by godbout** [comment]
> I'd love to see the ci", ci', ci[, etc motions!

yo!

they're implemented through the Accessibility Strategy, which means for apps where the text is readable/writable. when the text is not readable and/nor writable, kV cannot determine exactly the motion boundaries so it has to go through key remapping instead, and in those cases those moves are not doable. 

* the current kV implementation: https://github.com/godbout/kindaVim.docs
* the Accessibility Strategy implemented motions: https://github.com/godbout/kindaVim.docs/blob/master/AccessibilityStrategyImplementation.md#accessibility-strategy
* the Keyboard Strategy implemented motions: https://github.com/godbout/kindaVim.docs/blob/master/KeyboardStrategyImplementation.md#keyboard-strategy

you'll notice that ciB, cib etc. are also implemented (and also diB, viB, blah blah blah). 

hope this clarifies. lemme know if you need more details or explanation. i'll be glad to help.

**Comment by godbout** [comment]
> Thank you for clarifying! I originally misread the very helpful images, interpteting the two red X's to mean not implemented, but now I see those are just indicating count and repeat are not implemented for those

yeah, i see where the confusion may come from. only the motions implemented are listed yeah (else there would be too many). but i get you yeah, coming there on the first time it may not be that clear. hmm. may need to think of something to change.

> (which doesn't make sense for these motions anyway). Sorry for my confusion.

for which motion in particular are you talking about? `ci'` etc. have counts in Vim. same for `ci[` etc. so not sure which one exactly you're talking about.

**Comment by prurph** [comment]
> for which motion in particular are you talking about? `ci'` etc. have counts in Vim. same for `ci[` etc. so not sure which one exactly you're talking about.

This is news to me—I have never used count with any of the `i(` or `aB` motions and I didn't realize you could do that in vim! 😎

I certainly won't complain that you can't use a count with it in kindaVim though; the singular versions work great and are more than enough for me.

**Comment by godbout** [comment]
> This is news to me—I have never used count with any of the `i(` or `aB` motions and I didn't realize you could do that in vim! 😎

learning every day 😎️

> I certainly won't complain that you can't use a count with it in kindaVim though; the singular versions work great and are more than enough for me.

yeah. so the idea is that i build kV for myself 🤣️🤣️🤣️ if i don't need the move, i don't implement them at first. but if someone requests them, i do. hence the lack of count for those moves. enjoy!

**Comment by wookayin** [comment]
I't be great if all the basic vim motions can be supported. The following are a non-exhaustive list that doesn't work for me as of today:

- B, W, E: move backwards/forwards by WORD
- f<letter>, t<letter>
- "motion operators" (also request in the above) - viw, vi(, va{, ci(, etc.

**Comment by godbout** [comment]
> I use cib

😍😍😍

> and ciq

??? what motion is that lol. do you mean the quotes?

> are these motions supported? im trying on Legcord (modded Discord client) and it does not seem to work on it. Thanks in advance

yes they are supported. BUT. you can guess that to be able to calculate what's "within a block" or "within quotes", the app you're on needs to provide data (through the macOS Accessibility Framework). from what i see Legcord seems to be an Electron app, which means it uses Chromium, which means they don't provide proper AX data (been dealing with the Chromium devs to get their stuff fixed for the past 5 years, last [bug report last June](https://issues.chromium.org/423048796)). so because kV can't calculate and do text manipulations, it needs to revert to key remapping. when key remapping is involved, some moves can't be done. like `cib`, `ci"` etc.

for more info, here are the list of motions available:
1. for apps that support the macOS Accessibility properly: https://docs.kindavim.app/implementation/accessibility-strategy
2. for apps (like Electron) that don't support the macOS Accessibility, or send wrong data: https://docs.kindavim.app/implementation/keyboard-strategy

btw i'll add Legcord to The Wizard for the next release. for now you better add it to the `Key Mapping` Family so that kV is consistent. Electron apps return wrong data most of the time, but in some cases (depending on the number of lines, and you caret location), they by accident will return right data lol, so sometimes kV will use the Accessibility Strategy, sometimes (most of the time) the Keyboard Strategy. for consistency you better just enforce the Keyboard Strategy (again, by adding Legcord to the Key Mapping Family). best if that if you feel something doesn't work smoothly, you send the app to The Wizard and you see what it tells you! hope that clarifies.

**Comment by why-trv** [comment]
Maybe some `vim-surround` motions? Stuff like `ds'`, `cs"'`, `ysiw)` etc.
(https://github.com/tpope/vim-surround or https://github.com/kylechui/nvim-surround)

**Comment by godbout** [comment]
> Maybe some `vim-surround` motions? Stuff like `ds'`, `cs"'`, `ysiw)` etc. (https://github.com/tpope/vim-surround or https://github.com/kylechui/nvim-surround)

yes. the next big thing for kV is plugins. will update the ROADMAP when things settle down a bit, after the next release. should be in the following days.

**Comment by godbout** [comment]
(Surround request: https://github.com/godbout/kindaVim.blahblah/issues/150)

**Comment by RhvdG** [comment]
The sentence text object needs some love. I miss das, cas, cis, ...

**Comment by AryanRogye** [comment]
Hi! I just discovered Kindavim and it’s awesome.

The main reason I started looking for a tool like this is because I use Notion a lot, but it doesn’t have Vim support. One shortcut I rely on heavily is `^q`.

I recorded a small demo showing how this behaves in Neovim and how I’d love it to work with Kindavim:

https://github.com/user-attachments/assets/69657094-10f3-4438-a14f-bd518a0a6a55

It would be amazing if something like this could be supported.

**Comment by godbout** [comment]
> The main reason I started looking for a tool like this is because I use Notion a lot, but it doesn’t have Vim support. One shortcut I rely on heavily is `^q`.

yeah Visual Block. can't be done with the macOS AX APIs, so kV can't do it. you need to own your own buffer to be able to do that, which means each apps has to implement this themselves lol. that why we can't highlight all the matches in an incremental search too. the AX APIs allow you to have one continuous selection only, not multiple.

that being said it's "still" in my todo. but very very very low. maybe there's a completely different way to handle the problem. for example for incremental search i was thinking that _maybe_ kV could use OCR. but for your case of the Visual Block it's a totally different story because it's not just about visually highlighting the thing. so yeah. chance of this being possible to do is about 0.1% percent i think. chance of me digging it again/more, 100%. ETA, probably need to count in years tho because very low priority knowing that most probably, even if possible, that requires some heavy changes, different approach, whole new engines, etc.

**Comment by tannerkovach** [comment]
It seems '^' jumps to the first character in the line in normal mode but 'd^' doesn't work in operator mode, is there a technical reason for that or am I just missing something?

**Comment by godbout** [comment]
> It seems '^' jumps to the first character in the line in normal mode but 'd^' doesn't work in operator mode, is there a technical reason for that or am I just missing something?

good catch. just checked and yeah it's not implemented. no particular reason i think, most probably because i prioritize the moves i use myself and let users prioritize the other ones and `d^` never came up lol. `^` and `g^` are already done so mixing with the operators should be rather straightforward (just gonna require a lot of automated tests done). so pretty sure i can add `d^` (and its variants `c^`, `y^`, `dg^`, etc.) for the next release. will keep you updated.

**Comment by tannerkovach** [comment]
@godbout Wow thanks for the quick response! Love the project by the way, exactly what I was looking for and its super helpful. Great work! For the time being I setup karabiner-elements to map 'd^' to 'd0' for the time being as its close enough.

**Comment by godbout** [comment]
> [@godbout](https://github.com/godbout) Wow thanks for the quick response! Love the project by the way, exactly what I was looking for and its super helpful. Great work! For the time being I setup karabiner-elements to map 'd^' to 'd0' for the time being as its close enough.

my pleasure. this is gonna be fun. and always good to add new motions to kV. makes the whole thing better.

now you make me wonder tho, would there be a way in KE to get `d^` to work temporarily? 🤔 will implement native in kV for sure, but curious. i think i saw some stuff like almost you can move the caret to some sentence structure etc. or maybe i'm just high and have no idea what i'm talking about.

**Comment by godbout** [comment]
`d^`, `c^`, `y^`, `dg^`, `cg^`, `yg^` tests and impl done for kV Engine that is dispatching the motions to the proper Strategies. still tests and impl to be done for at least the Accessibility Strategy, all moves. yeah i'm gonna update the status here for my own self 🤣

**Comment by godbout** [comment]
> `d^`, `c^`, `y^`, `dg^`, `cg^`, `yg^` tests and impl done for kV Engine that is dispatching the motions to the proper Strategies. still tests and impl to be done for at least the Accessibility Strategy, all moves. yeah i'm gonna update the status here for my own self 🤣

all moves tests and impl done for the kV Engine dispatching regarding the Keyboard Strategy too. moving to implementing the core of the moves now.

**Comment by godbout** [comment]
tests and impl done for the Keyboard Strategy (Key Mapping and 911 Families). next is implementing all the moves for the Accessibility Strategy.

**Comment by godbout** [comment]
tests and impl done for the Accessibility Strategy, `c^`, `cg^`, `y^`, `yg^`. last ones to do `d^` and `dg^`. planning a release for next week.

**Comment by godbout** [comment]
@tannerkovach released in kV91: https://github.com/godbout/kindaVim.blahblah/releases/tag/91 all tested and all but please lemme know if you find cases that don't follow Vim's behavior. should be fine but you never. there's always some weird edge cases. may have forgot some.

**Comment by Yahddyyp** [comment]
I would like to request a motion for when you press escape in using a sequence(jj), vim mode for it go back into inset mode like regular vim behaviour

**Comment by godbout** [comment]
> I would like to request a motion for when you press escape in vim mode for it go back into inset mode like regular vim behaviour

? kV follows Vim on this. `esc` doesn't go back to Insert Mode. or am i misunderstanding something?

**Comment by Yahddyyp** [comment]
i mean when i use a sequence like jj or something or than that it does not allow me to enter insert mode again unless i press a or i

**Comment by godbout** [comment]
> i mean when i use a sequence like jj or something or than that it does not allow me to enter insert mode again unless i press a or i

yeah, that's Vim behavior.

see:
* actually developing the opposite of what you want, which is the correct Vim behavior: https://github.com/godbout/kindaVim.blahblah/issues/206
* using sequence and Karabiner-Elements integration to do what you want: https://github.com/godbout/kindaVim.blahblah/issues/125
* another example of using Karabiner-Elements to do what you want: https://github.com/godbout/kindaVim.blahblah/issues/272

so basically kV follows Vim. but you can still do what you want using the custom mapping with KE integration.

**Comment by dkarter** [comment]
hi! i'm really missing `c` when using visual mode, so i select a word and hit `c` - i expect it to delete the word and put the cursor in insert where it was - similar to what vim does

**Comment by godbout** [comment]
> hi! i'm really missing `c` when using visual mode, so i select a word and hit `c` - i expect it to delete the word and put the cursor in insert where it was - similar to what vim does

https://github.com/user-attachments/assets/395e1763-4148-42a2-9800-24870db9025a

definitely there. also in the lists at https://docs.kindavim.app/implementation/accessibility-strategy#visual-mode and https://docs.kindavim.app/implementation/keyboard-strategy#visual-mode (both `c` and `s`).

so or you're using kV on an app that should be in the PG-R Family but isn't put there yet, or you're using a very stubborn app 😂 on which app is this happening?
